Migrate ensure_executing transitions to new WorkerState event mechanism - part 2

crusaderky opened the pull request:
- Partially closes #5895

### In scope for this PR
- Migrate ``def ensure_computing() -> None`` to ``def _ensure_computing() -> RecsInstrs``

### Out of scope for this PR; in scope for the issue 
- Implement ``Worker.stimulus_history``
- Track tasks spawned by ``Worker._handle_instructions`` and cancel them in Worker.close

### Out of scope for the issue
- Send Event Python objects from the Scheduler
- Other events (replicate etc.)

crusaderky commented on distributed/tests/test_steal.py:
```diff
@@ -1182,17 +1182,20 @@ async def test_reschedule_concurrent_requests_deadlock(c, s, *workers):
+        ("already-computing", victim_key, "executing", wsB.address, wsC.address),
+        ("already-aborted", victim_key, "executing", wsA.address),
+    ]
+    assert msgs in (expect1, expect2)
```
Before this PR, ``victim_ts.who_has != {wsC}`` was *never* true. After my change, it is *always* true. I don't think I fully understand what happened here, and if this is OK or it warrants further investigation.

- CC @fjetter 
- XREF https://github.com/dask/distributed/issues/5370

fjetter replied:
I encountered a third possibility with `("already-aborted", victim_key, "memory", wsA.address)`

fjetter replied:
We should move the event setting and fut gathering to the bottom. the steal responses are already logged at this point in time such that we can assert them. We only set the event and collect the futures to allow the threadpool to properly shut down but the test actually doesn't require them to finish

fjetter replied:
This is an interesting change, indeed. It is expected.

Before, the key was _sometimes_ in state `ready`, other times in state `executing` when the steal request (B->C) reaches the workers. I don't believe that on `main` this condition is never true, rather it is unlikely.
Whether or not the task was in state ready or executing would depend on whether or not something called `ensure_computing` before the steal request could've been handled, e.g. if any other message comes over the stream before the steal request, we might've transitioned the key from ready to executing since every stream message triggers an ensure_compute.

This PR removes this ambiguity by ensuring that a task is transitioned to executing in the very same tick as it was transitioned to ready. It removes the possibility of this race condition entirely. That's great 🎉 
The ambiguity that's still in the test (expected1 vs expected2) is outside of our control since this depends on the which workers responds faster which is almost impossible to control.

we should, however, move the `ev.set()` to the bottom to remove the third psosibiltiy of the task already done computing

crusaderky replied:
If you move ev.set() to the bottom, the story is empty

crusaderky replied:
> I encountered a third possibility with `("already-aborted", victim_key, "memory", wsA.address)`

Is it
```
    expect3 = [
        ("already-computing", victim_key, "executing", wsB.address, wsC.address),
        ("already-aborted", victim_key, "memory", wsA.address),
    ]
```
or
```
    expect3 = [
        ("already-aborted", victim_key, "memory", wsA.address),
    ]
```
?

crusaderky commented on distributed/worker.py:
```diff
@@ -2038,9 +2040,10 @@ def transition_executing_rescheduled(
+        return merge_recs_instructions(
+            ({ts: "released"}, [RescheduleMsg(key=ts.key, worker=self.address)]),
+            self._ensure_computing(),
+        )
```
I did this in all transitions that invoke ``self._executing.discard(...)``

crusaderky commented on distributed/worker.py:
```diff
@@ -2347,7 +2357,6 @@ def transition_ready_executing(
             )
 
         ts.state = "executing"
-        self._executing.add(ts)
```
This line has moved to _ensure_computing. I'm not happy about it but I could not find a cleaner solution.

crusaderky commented on distributed/worker.py:
```diff
@@ -3370,41 +3371,60 @@ async def _maybe_deserialize_task(
+                assert ts not in self._executing
+
+            recs[ts] = "executing"
+            self._executing.add(ts)
```
Moved from ``transition_constrained_executing`` and ``transition_ready_executing``. I'm not happy about it but I could not find a cleaner solution.

crusaderky commented on distributed/worker.py:
```diff
@@ -3538,6 +3558,23 @@ async def execute(self, key: str, *, stimulus_id: str) -> StateMachineEvent | No
+        This is a temporary hack to be removed as part of
+        https://github.com/dask/distributed/issues/5894.
+        """
+        return self._ensure_computing()
```
Called through ``self.ensure_computing`` from

- ``Worker.handle_scheduler``
- ``Worker.gather_dep``
- ``Worker.find_missing``
- ``Worker.release_key``

crusaderky commented:
Merged all blockers. This is ready for final review and merge.

fjetter commented on distributed/tests/test_worker_state_machine.py:
```diff
@@ -109,3 +111,17 @@ def test_sendmsg_to_dict():
     assert smsg.to_dict() == {"op": "release-worker-data", "key": "x"}
+
+
+def test_merge_recs_instructions():
```
What is the behaviour of `merge_recs_instructions` if a key in recommendations is duplicated?

crusaderky replied:
The newest one silently wins. I'm changing it to fail if one tries to change a previous recommendation with a new, non-identical one. This has a small performance cost which however is probably worth having.

fjetter replied:
I don't care too much which way this goes but think we should be explicit about the behaviour. Being strict is fine but I've written code that deliberately overwrote a recommendation so I could see valid use cases in both directions.

If we get it working being strict, I'm happy

fjetter reviewed (commented):
It should be possible to remove the ensure_computing handler entirely. At least locally, the tests pass for me

```diff
diff --git a/distributed/worker.py b/distributed/worker.py
index 96d95373..df8fe680 100644
--- a/distributed/worker.py
+++ b/distributed/worker.py
@@ -110,7 +110,6 @@ from distributed.worker_state_machine import (
     AddKeysMsg,
     AlreadyCancelledEvent,
     CancelComputeEvent,
-    EnsureComputingEvent,
     Execute,
     ExecuteFailureEvent,
     ExecuteSuccessEvent,
@@ -1181,7 +1180,7 @@ class Worker(ServerNode):
     async def handle_scheduler(self, comm):
         try:
             await self.handle_stream(
-                comm, every_cycle=[self.ensure_communicating, self.ensure_computing]
+                comm, every_cycle=[self.ensure_communicating]
             )
         except Exception as e:
             logger.exception(e)
@@ -2024,7 +2023,7 @@ class Worker(ServerNode):
             assert ts.key not in self.ready
         ts.state = "constrained"
         self.constrained.append(ts.key)
-        return {}, []
+        return self._ensure_computing()
 
     def transition_long_running_rescheduled(
         self, ts: TaskState, *, stimulus_id: str
@@ -3044,7 +3043,6 @@ class Worker(ServerNode):
                         recommendations[ts] = "fetch" if ts.who_has else "missing"
                 del data, response
                 self.transitions(recommendations, stimulus_id=stimulus_id)
-                self.ensure_computing()
 
                 if not busy:
                     self.repetitively_busy = 0
@@ -3085,7 +3083,6 @@ class Worker(ServerNode):
                     "find-missing"
                 ].callback_time = self.periodic_callbacks["heartbeat"].callback_time
                 self.ensure_communicating()
-                self.ensure_computing()
 
     async def query_who_has(self, *deps: str) -> dict[str, Collection[str]]:
         with log_errors():
@@ -3218,7 +3215,6 @@ class Worker(ServerNode):
 
             self._executing.discard(ts)
             self._in_flight_tasks.discard(ts)
-            self.ensure_computing()
             self.ensure_communicating()
 
             self._notify_plugins(
@@ -3370,11 +3366,6 @@ class Worker(ServerNode):
             )
             raise
 
-    def ensure_computing(self) -> None:
-        self.handle_stimulus(
-            EnsureComputingEvent(stimulus_id=f"ensure_computing-{time()}")
-        )
-
     def _ensure_computing(self) -> RecsInstrs:
         if self.status in (Status.paused, Status.closing_gracefully):
             return {}, []
@@ -3558,14 +3549,6 @@ class Worker(ServerNode):
     def handle_event(self, ev: StateMachineEvent) -> RecsInstrs:
         raise TypeError(ev)  # pragma: nocover
 
-    @handle_event.register
-    def _(self, ev: EnsureComputingEvent) -> RecsInstrs:
-        """Let various methods of worker give an artificial 'kick' to _ensure_computing.
-        This is a temporary hack to be removed as part of
-        https://github.com/dask/distributed/issues/5894.
-        """
-        return self._ensure_computing()
-
     @handle_event.register
     def _(self, ev: UnpauseEvent) -> RecsInstrs:
         """Emerge from paused status. Do not send this event directly. Instead, just set
diff --git a/distributed/worker_state_machine.py b/distributed/worker_state_machine.py
index 51c5a768..5c07fced 100644
--- a/distributed/worker_state_machine.py
+++ b/distributed/worker_state_machine.py
@@ -358,17 +358,6 @@ class StateMachineEvent:
     __slots__ = ("stimulus_id",)
     stimulus_id: str
 
-
-@dataclass
-class EnsureComputingEvent(StateMachineEvent):
-    """Let various methods of worker give an artificial 'kick' to _ensure_computing.
-    This is a temporary hack to be removed as part of
-    https://github.com/dask/distributed/issues/5894.
-    """
-
-    __slots__ = ()
-
-
 @dataclass
 class UnpauseEvent(StateMachineEvent):
     __slots__ = ()

```

crusaderky commented:
All review comments incorporated; let's see if it remains green

fjetter commented on distributed/worker.py:
```diff
@@ -3370,41 +3371,60 @@ async def _maybe_deserialize_task(
+
+            self.constrained.popleft()
+            for resource, needed in ts.resource_restrictions.items():
+                self.available_resources[resource] -= needed
```
FWIW I agree with you and am not very happy about it either. In a perfect world, `_ensure_computing` would be idempotent but hardly feasible with the heaps. 

Regarding the _executing and/or available resources, we _could_ work around this by establishing counters at the beginning of the body. Not sure if this is worth it, though.

The way I can reason that this still makes sense is that our invariants (i.e. everything we validate in `self.validate_state` must be correct only after a stimulus ID finished all their transitions but intermediate states may violate certain constraints.

fjetter commented:
Most failures are unrelated and some are already fixed on main

`distributed/tests/test_priorities.py::test_persist` is failing often and I haven't seen this fail before. I'll have a look

crusaderky commented:
This PR is now blocked by and incorporates #6077

crusaderky commented:
I think I found an important design problem.

Patch to #6077:
```patch
--- a/distributed/worker.py
+++ b/distributed/worker.py
@@ -1831,6 +1831,7 @@ class Worker(ServerNode):
         annotations: dict | None = None,
         stimulus_id: str,
     ) -> None:
+        print(f"handle_compute_task({key}); {self.tasks=}")
         self.log.append((key, "compute-task", stimulus_id, time()))
         try:
             ts = self.tasks[key]
@@ -3372,6 +3373,8 @@ class Worker(ServerNode):
     def ensure_computing(self) -> None:
         if self.status in (Status.paused, Status.closing_gracefully):
             return
+        if self.ready and self.executing_count < self.nthreads:
+            print(f"ensure_computing: ready={sorted(self.ready)} {self.tasks=}")
         try:
             stimulus_id = f"ensure-computing-{time()}"
             while self.constrained and self.executing_count < self.nthreads:
```
Patch to this PR:
```patch
--- a/distributed/worker.py
+++ b/distributed/worker.py
@@ -1830,6 +1830,7 @@ class Worker(ServerNode):
         annotations: dict | None = None,
         stimulus_id: str,
     ) -> None:
+        print(f"handle_compute_task({key}); {self.tasks=}")
         self.log.append((key, "compute-task", stimulus_id, time()))
         try:
             ts = self.tasks[key]
@@ -3367,7 +3368,8 @@ class Worker(ServerNode):
     def _ensure_computing(self) -> RecsInstrs:
         if self.status in (Status.paused, Status.closing_gracefully):
             return {}, []
-
+        if self.ready and self.executing_count < self.nthreads:
+            print(f"ensure_computing: ready={sorted(self.ready)} {self.tasks=}")
         recs: Recs = {}
         while self.constrained and len(self._executing) < self.nthreads:
             key = self.constrained[0]
```

Output of test_priorities::test_compute():
#6077:
```
handle_compute_task(low); self.tasks={}
handle_compute_task(clog); self.tasks={'low': <TaskState 'low' ready>}
handle_compute_task(high); self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>}
ensure_computing: ready=[((-1, 1, 0, -2), 'high'), ((0, 1, 0, -1), 'clog'), ((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>, 'high': <TaskState 'high' ready>}
ensure_computing: ready=[((0, 1, 0, -1), 'clog'), ((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>, 'high': <TaskState 'high' memory>}
handle_compute_task(low); self.tasks={}
handle_compute_task(clog); self.tasks={'low': <TaskState 'low' ready>}
handle_compute_task(high); self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>}
ensure_computing: ready=[((-1, 1, 0, -2), 'high'), ((0, 1, 0, -1), 'clog'), ((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>, 'high': <TaskState 'high' ready>}
ensure_computing: ready=[((0, 1, 0, -1), 'clog'), ((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>, 'high': <TaskState 'high' memory>}
ensure_computing: ready=[((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' memory>, 'high': <TaskState 'high' memory>}
ensure_computing: ready=[((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' memory>, 'high': <TaskState 'high' memory>}
```
This PR:
```
handle_compute_task(clog); self.tasks={}
ensure_computing: ready=[((0, 1, 0, 0), 'clog')] self.tasks={'clog': <TaskState 'clog' ready>}
handle_compute_task(high); self.tasks={'clog': <TaskState 'clog' executing>}
handle_compute_task(low); self.tasks={'clog': <TaskState 'clog' executing>, 'high': <TaskState 'high' ready>}
handle_compute_task(clog); self.tasks={}
ensure_computing: ready=[((0, 1, 0, 0), 'clog')] self.tasks={'clog': <TaskState 'clog' ready>}
handle_compute_task(high); self.tasks={'clog': <TaskState 'clog' executing>}
handle_compute_task(low); self.tasks={'clog': <TaskState 'clog' executing>, 'high': <TaskState 'high' ready>}
Dumped cluster state to test_cluster_dump/test_compute.yaml
ensure_computing: ready=[((-1, 1, 0, -1), 'high'), ((1, 1, 0, -2), 'low')] self.tasks={'clog': <TaskState 'clog' error>, 'high': <TaskState 'high' ready>, 'low': <TaskState 'low' ready>}
```

In both branches, the order of the first three handle_compute_task calls is random and changes at every run.
I verified that this matches the order of the calls to ``scheduler.py::_task_to_msg`` on the scheduler side.

In main, ensure_computing is not executed until after all three calls to handle_compute_task. Hence, their order does not matter and ensure_computing finds a complete ``ready`` heap and can deterministically pick up the task with the lowest priority (high).

In this PR, ensure_computing is executed immediately after the first call to handle_compute (which sounds a lot more robust to me), but the obvious consequence is that the ready heap is incomplete and just contains a random task in it when this happens.

The correct solution IMHO is to ensure the scheduler invokes handle_compute_task in the correct order to begin with?

# Side note 1
I can't understand why there are more than 3 calls to handle_compute_task and why the task states are flip-flopping from memory back to ready.

# Side note 2
From this I reason that, as soon as you exceed the chunk size for batched sends, then priority should go out of the window.
However I have proven this is not true and I have no idea about why.

This is green in main and I can't figure out how it's possible:
```python

@pytest.mark.parametrize("ntasks", [10, 100, 1000, 2000, 4000])
@gen_cluster(client=True, nthreads=[])
async def test_large_batches(c, s, ntasks):
    """Test that the scheduling order is unperturbed by the chunk size of the batched
    comms
    """
    futs = [c.submit(inc, key=str(i), priority=i) for i in range(ntasks)]
    while len(s.tasks) < ntasks:
        await asyncio.sleep(0.01)

    async with Worker(s.address, nthreads=1) as a:  # See comment in test_submit
        await wait(futs)
        story = a.story(*(str(i) for i in range(ntasks)))

    story = [
        int(ev[0])
        for ev in story
        if ev[1:3] == ("ready", "executing")
    ]
    assert story == list(reversed(range(ntasks)))
```

fjetter commented:
Replacing this set with a heap should not be a big deal. However, there is one place where we're discarding a task and this is not implemented on our heap (UniqueTaskHeap)

fjetter commented:
Is this test breaking or is this just something you noticed during debugging? If the test is OK I would suggest to file an issue about this and move on for now

fjetter commented:
A probably simpler approach is to just sort the set once by priority. This might be costly but I don't expect this to happen very frequently 
https://github.com/dask/distributed/blob/8cf11967ca4c159c23f328b2dc5b2839e8c46383/distributed/scheduler.py#L4586

fjetter commented:
All "ordinary" task assignments and priorities should be handled properly by [this sort](https://github.com/dask/distributed/blob/8cf11967ca4c159c23f328b2dc5b2839e8c46383/distributed/scheduler.py#L4942-L4944) so I assume this is just us leaving the unrunnable state. I think sorting should be the way to go there

fjetter commented:
I believe this prioritization issue is why https://github.com/dask/distributed/pull/5443 does not work

crusaderky commented:
All green! Started tests one last time.
Ready for final review and merge

fjetter commented on distributed/tests/test_steal.py:
```diff
@@ -1182,17 +1182,25 @@ async def test_reschedule_concurrent_requests_deadlock(c, s, *workers):
+        ("already-computing", victim_key, "executing", wsB.address, wsC.address),
+        ("already-aborted", victim_key, "executing", wsA.address),
+    ]
+    # This outcome appears only in ~2% of the runs
+    expect3 = [
+        ("already-computing", victim_key, "executing", wsB.address, wsC.address),
+        ("already-aborted", victim_key, "memory", wsA.address),
+    ]
```
This can be avoided by moving the ev.set (and gather) after the stories are collected

crusaderky replied:
If you do that, the story is completely empty.

fjetter reviewed (approved):
I'm curious about the additional `ensure_communicating`. It should not be a big deal but I'm wondering if this accidentally slipped in or if there is a reason.

Otherwise LGTM

crusaderky commented on distributed/scheduler.py:
```diff
@@ -3611,6 +3611,24 @@ def _reevaluate_occupancy_worker(self, ws: WorkerState):
                 steal.recalculate_cost(ts)
 
+    @ccall
+    def bulk_schedule_after_adding_worker(self, ws: WorkerState):
```
where?

crusaderky commented on distributed/worker.py:
```diff
@@ -3209,6 +3213,7 @@ def release_key(
 
             self._executing.discard(ts)
             self._in_flight_tasks.discard(ts)
+            self.ensure_communicating()
```
There was a change in _in_flight_task, so I thought it would be necessary to prod the comms.
But I guess it can be removed.

fjetter replied:
I don't think this has an impact. It shouldn't be harmful, though. I was just curious

crusaderky commented:
I just realised that release_key should never be called on an executing task.
If it could, we would end up with a stuck worker.
See my latest commit to address the issue.

fjetter reviewed (approved):
The last commit is indeed _delicate_. However, the way how the transition system is designed, i.e. by having the cancelled state, it is **impossible** to arrive in a condition where `release_key` is called on a task that is still executing.

Code coverage, however flaky it is, also show that these lines are indeed never hit.

For a similar argument, ts should also never be in _in_flight_tasks but let's not push our luck; I'd prefer a dedicated PR for changes like this

crusaderky commented on distributed/worker.py:
```diff
@@ -3181,26 +3184,20 @@ def release_key(
+
+            try:
+                self.data.pop(key, None)
+            except FileNotFoundError:
+                logger.error("Tried to delete %s but no file found", exc_info=True)
```
Yes I spotted that too. I agree this should not be here.

crusaderky commented:
- The previous change to release_key caused failures. I reverted it and replaced it with https://github.com/dask/distributed/pull/6062/commits/f22099d329f073eece47e59bf3d7c1faad604d54
- I spotted a test failure (can't find it anymore) due a duplicate in Worker.ready. I have no idea how to reproduce the issue. For now I'm dealing with it gracefully: https://github.com/dask/distributed/pull/6062/commits/f53480ccb7d8f419ad3315ec8810e9fccb2da90b

fjetter commented:
Are the changes to release_key really necessary? This is something I would much rather do in a dedicated PR

fjetter commented:
> I spotted a test failure (can't find it anymore) due a duplicate in Worker.ready. I have no idea how to reproduce the issue. For now I'm dealing with it gracefully:

I think I would prefer us to fail rather than ignore the issue, otherwise we will never have the chance to fix it. In the past we've had many, many of these "let's just ignore that this task should be in a different state" which ultimately caused a lot of inconsistencies

fjetter commented:
`self.ready` can have duplicates and that’s almost unavoidable.

That basically happens if
- a task is assigned to a Worker and transitioned to ready (heappush)
- The task is stolen (no way to pop from heap, the task stays there)
- The task is assigned to the worker again (heappush again)
This is why there are the guards in place to check whether a task is in fact still in state ready https://github.com/dask/distributed/blob/a40f20553a9f6edd4e7072239744d71c6e0c0bcb/distributed/worker.py#L3399-L3400

fjetter commented on distributed/worker.py:
```diff
@@ -3369,42 +3366,61 @@ async def _maybe_deserialize_task(
+                self.constrained.popleft()
+                continue
+
+            # FIXME We should never have duplicates in self.constrained or self.ready;
+            #       however replacing the block below and the matching ones later in
+            #       this function for the ready queue with just 'assert ts not in recs'
+            #       causes *sporadic* failures in the test suite.
```
See https://github.com/dask/distributed/pull/6062#issuecomment-1093103470
 Unless we're willing to pay for a delete-from-heap or disable stealing, we need to allow duplicates.
We have similar problems with, for example, `pending_data_per_worker` where I introduced the `UniqueTaskHeap`

fjetter commented on distributed/worker.py:
```diff
@@ -2254,12 +2261,12 @@ def transition_cancelled_released(
+        return merge_recs_instructions(
+            self.transition_generic_released(ts, stimulus_id=stimulus_id),
+            ({ts: next_state} if next_state != "released" else {}, []),
+            self._ensure_computing(),
```
If we call `_ensure_computing` as part of `transition_generic_released` we no longer need this here, do we?

fjetter replied:
I assume calling it twice is harmless. I'm OK with keeping it. I actually don't like it being part of `transition_generic_released` but there is probably more work hidden so I'm good

crusaderky replied:
removed

fjetter reviewed (approved):
Third and hopefully last ✅ if tests are cool with it

Great job @crusaderky ! I believe this is a big step towards a cleaner version of the worker even though it might still feel messy
